tests/client/factory_test.py

```diff
@@ -455,6 +486,128 @@ def test_ssl_error(self, es):
+        if version == "7.9.0":
+            calls = [
+                mock.call({"id": "baz"}),
+                mock.call({"id": "bar"}),
+                mock.call({"id": "foo"}),
+            ]
+        else:
+            calls = [mock.call({"ids": ids})]
+
+        assert client.delete_api_keys(es, ids, max_attempts=3)
+        assert es.security.invalidate_api_key.call_args_list == calls
+
+    @pytest.mark.parametrize("version", ["7.9.0", "7.10.0"])
+    @mock.patch("time.sleep")
+    @mock.patch("elasticsearch.Elasticsearch")
+    def test_retries_api_keys_deletion_on_transport_errors(self, es, sleep, version):
+        max_attempts = 5
+        es.info.return_value = {"version": {"number": version}}
+        ids = ["foo", "bar", "baz"]
+        if version == "7.9.0":
+            es.security.invalidate_api_key.side_effect = [
+                {"invalidated_api_keys": ["baz"]},
+                {"invalidated_api_keys": ["bar"]},
+                elasticsearch.TransportError(503, "Service Unavailable"),
+                elasticsearch.TransportError(401, "Unauthorized"),
+                {"invalidated_api_keys": ["foo"]},
+            ]
+            calls = [
+                # baz and bar are deleted successfully, leaving only foo
+                mock.call({"id": "baz"}),
+                mock.call({"id": "bar"}),
+                # two exceptions are thrown, so it should take 3 attempts to delete foo
+                mock.call({"id": "foo"}),
+                mock.call({"id": "foo"}),
+                mock.call({"id": "foo"}),
+            ]
+        else:
+            es.security.invalidate_api_key.side_effect = [
+                elasticsearch.TransportError(503, "Service Unavailable"),
+                elasticsearch.TransportError(401, "Unauthorized"),
+                elasticsearch.TransportError(408, "Timed Out"),
+                elasticsearch.TransportError(408, "Timed Out"),
+                {"invalidated_api_keys": ["foo", "bar", "baz"]},
+            ]
+            calls = [mock.call({"ids": ids}) for _ in range(max_attempts)]
+
+        assert client.delete_api_keys(es, ids, max_attempts=max_attempts)
+        assert es.security.invalidate_api_key.call_args_list == calls
+
+    @pytest.mark.parametrize("version", ["7.9.0", "7.10.0"])
+    @mock.patch("elasticsearch.Elasticsearch")
+    def test_raises_exception_when_api_key_deletion_fails(self, es, version):
+        es.info.return_value = {"version": {"number": version}}
+        ids = ["foo", "bar", "baz"]
+        es.security.invalidate_api_key.side_effect = [
+            elasticsearch.TransportError(503, "Service Unavailable"),
+            elasticsearch.TransportError(401, "Unauthorized"),
+            Exception("Whoops!"),
+        ]
+
+        with pytest.raises(exceptions.RallyError, match=re.escape(f"Could not delete API keys with the following IDs: {ids}")):
+            client.delete_api_keys(es, ids)
+
```

Comment:
This actually raised a more substantial issue in my mind, which I've addressed in https://github.com/elastic/rally/pull/1520/commits/8eec4b167f934fff73b59e56c6ba9f234503b0f9:

It's possible for the 7.10.0+ version of the deletion code to fail silently if we rely just on exceptions to catch errors. This is because it's basically a bulk request, which means that the response can contain both successful and unsuccessful deletions but still have an HTTP 200 status code. That commit handles that scenario, modifies the logic for the "legacy" code, and refactors the relevant tests.